bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/service/ReadyServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.internal.service;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+
+import org.eclipse.smarthome.core.service.ReadyMarker;
+import org.eclipse.smarthome.core.service.ReadyMarkerFilter;
+import org.eclipse.smarthome.core.service.ReadyService;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the {@link ReadyService} interface.
+ *
+ * @author Simon Kaufmann - initial contribution and API.
+ *
+ */
+@Component
+public class ReadyServiceImpl implements ReadyService {
+
+    private static final Logger logger = LoggerFactory.getLogger(ReadyServiceImpl.class);
+    private static final ReadyMarkerFilter ANY = new ReadyMarkerFilter();
```

Comment:
Perhaps we could make the ANY constructor (that uses null twice) private to and use a static function that returns an instance (no need for a singleton). Perhaps the meaning of `ReadyMarkerFilter.any()` is more intuitive then `new ReadyMarkerFilter()`.

Reply:
If i make that private, there is no public constructor left. My intention was to keep the one with arguments private in order to prevent the constructor-mess we had at other places and rather have the builder-like pattern using `with...()`-Methods. 

So do I understand you right that using `ReadyMarkerFilter.any().withType("...");` is the usage you are suggesting?

Reply:
No :wink: 

If I understand the current situation correctly you can thee different type of ready markers. 

* filter type, created by `ReadyMarkerFilter.withType(String type)`
* filter identifier, created by `ReadyMarkerFilter.withIdentifier(String type)`
* filter nothing (the any one), created by `new ReadyMarkerFilter()`

Correct me if I am wrong.

So, I just think about making all constructors private and use for the last one a factory method, too. All three functions should return immutable objects.

* filter type, created by `ReadyMarkerFilter.withType(String type)`
* filter identifier, created by `ReadyMarkerFilter.withIdentifier(String type)`
* filter nothing (the any one), created by `ReadyMarkerFilter.any()`

Reply:
Ah, that's the misunderstanding... the `with...(...)`-Methods are not static. Therefore you can create:

* filter nothing, created by `new ReadyMarkerFilter()`
* filter type, created by `new ReadyMarkerFilter().withType(String type)`
* filter identifier, created by `new ReadyMarkerFilter().withIdentifier(String identifier)`
* filter for both, created by  `new ReadyMarkerFilter().withType(String type).withIdentifier(String identifier)`

The reasoning was to have a design is that it can be evolved/extended with further parameters without creating a mess with the string arguments in constructors. I'm really not sure it these two parameters will be the only ones needed in future.

Reply:
> the `with...(...)`-Methods are not static

It seems I have forgotten how to read...